library/src/scala/internal/quoted/PickledQuote.scala

```diff
@@ -0,0 +1,45 @@
+package scala.internal.quoted
+
+import scala.quoted._
+import scala.internal.tasty.CompilerInterface.quoteContextWithCompilerInterface
+
+/** Pickled representation of a quoted expression or type */
+trait PickledQuote:
+
+  /** Bytes of the TASTy containing the quoted expression or type */
+  def bytes(): Array[Byte]
```

Comment:
In the original design, the interface defines abstract methods such as `unpickle()`. Will it be more flexible to keep the original design without committing to a concrete representation in the interface?

Reply:
We have the same flexibility with both designs. In this one we just need to add can add a `PickledQuote.unpickleV2` if we need to change the unpickling logic without breaking the old one. Though `make` is the one that will most likely evolve as it contains the byte representation logic and the hole logic.

Reply:
The old `PickledExpr` interface was guided by an attempt to have a user-facing interface for pickled quotes. But we could provide the same abstraction and internally use `PickledQuote` to achieve the same.

Reply:
As we put the protocol methods in a single trait. What will happen if we put the methods in `CompilerInterface` similar to the original design before this PR? Will that have the same benefit?

```Scala
trait CompilerInterface {
  type PickledQuote
  type PickledArgs
  def unpickleExpr(repr: PickledQuote, args: PickledArgs): Term
  def unpickleType(repr: PickledQuote, args: PickledArgs): TypeTree
}
```

Reply:
How would one create a `PickledQuote` in that design? We need to create it with some concrete class that exists in the library, but that design seems to indicate that it is not known in the library.

Reply:
An abstract constructor `pickleExpr` and `pickleType` can be defined if needed:

```Scala
trait CompilerInterface {
  type PickledQuote
  type PickledArgs
  def pickleExpr(tree: Term): PickledQuote
  def pickleType(tree: TypeTree): PickledQuote
  def unpickleExpr(repr: PickledQuote, args: PickledArgs): Term
  def unpickleType(repr: PickledQuote, args: PickledArgs): TypeTree
}
```

It is just a reshuffling of the library code.

Reply:
That does not help. You would need a `Term` to create a `PickledQuote` and a `PickledQuote` to create `Term`. 
We need a way to create one from a serialized version that is encoded in the code of the app, which implies that one needs to use abstractions that are defined in the library.

Having a `pickleExpr` may be useful in other use-cases. For example, serializing a closed expression and sending it to another machine for it to execute. This should probably be implemented in staging only though.

Reply:
> We need a way to create one from a serialized version that is encoded in the code of the app, which implies that one needs to use abstractions that are defined in the library.

Maybe I miss something here. Isn't reflect interface already in the library? It's just a reshuffling of the library.

Reply:
The thing is that the implementation of the contents of `PickledQuote` for reified quotes must be implemented in the app (or library). The compiler cannot implement them in the implementation of the `CompilerInterface` it must generate a new implementation for each quote in the source.

Reply:
It seems to me that it is not a problem. Because at the place where the user code plays with quotes, there is a `refl: CompilerInterface` instance in scope, which has all the functionality.

Reply:
That is not the issue. The question is what code is generated for the pickled quote

Reply:
Can the generated code be something like the following:

```Scala
// assume `def qctx(using ctx: QuoteContext): ctx.type = ctx` in stdlib.

qctx.reflect.unpickle("....", args)
```

Reply:
But that goes back to having a single representation. That is basically the current version.

Reply:
How do you represent/create the args?

Reply:
> But that goes back to having a single representation. That is basically the current version.

It does not have to be like this. If there is a need to abstract representation detail, it suffices to introduce an abstract type.  The code example is just a reshuffling of the library code -- they are just different ways to implement abstraction.

Reply:
> How do you represent/create the args?

I might not fully understand the requirement. Will something like the following work:

```Scala
trait CompilerInterface {
  type PickledQuote
  def unpickleExpr(repr: PickledQuote, args: Seq[Tree]): Term
  def unpickleType(repr: PickledQuote, args: Seq[Tree]): TypeTree
  def pickleExpr(expr: Tree): PickledQuote
  def pickleType(tpt: Tree): PickledQuote
}
```

The API still provides the flexibility for the compiler to have different implementations of `PickledQuote`.

Reply:
The point is not to rely on the representation that is stated in the arguments of `pickleTasty`. This would make it less flexible. We would not be able to avoid the allocations of the seq and the lambdas. We could also not change the representation of the pickled tasty in the string. 

The whole point is to be able to change the implementation of the pickled quoted without having to change the interface.

Reply:
> The point is not to rely on the representation that is stated in the arguments of `pickleTasty`. This would make it less flexible. We would not be able to avoid the allocations of the seq and the lambdas. We could also not change the representation of the pickled tasty in the string.
> 

It seems to me it is not different in essence from defining the following method in stdlib:

```Scala
  def make(pickled: List[String], seq: Seq[Seq[Any /* Expr[Any] | Type[?] */] => Any]): PickledQuote = ...
```

It seems it's just doing the same thing, in a zig-zag way:

- If it's possible to change implementation of `make` without change the interface, then the same can be done to the implementation of `pickleTasty`
- If it's possible to change the signature of `make`, then make the same change to `pickleTasty`.
- If it is intended to create another `make2`, it is possible to add `pickle2`, etc.

Reply:
It can be done but changes would be harder. And we still need an implementation for each one of those on the library side as is the current case.

Reply:
@liufengyun can I merge this one? I will start working on improvements of the code generation after this. That would work as a test to see how easy it is to evolve code with this interface. We can always change it later if it does not work.

Reply:
Just repeat the message in the email, it seems I miss something about the following point:

> And we still need an implementation for each one of those on the library side as is the current case.

I think `qctx.reflect.pickTasty(...)` will remove the need to put concrete implementation in stdlib.

Reply:
Indeed, it could all be implemented on the compiler. But that means that each improvement will be painfully complex in the compiler (i.e. the reason why we have not been able to improve it so far) and would have to add a new method on the interface on each change. I don't see how that would scale. The version I propose has the option of adding new static helper method if it is simpler or lets the code be generated without the CompilerInterface which is more flexible.

If we do it your way we also need to add the static helper methods as we do not have access to `pickTasty` method directly because we don't have access to the `qctx` where the transformation happens.

Reply:
> If we do it your way we also need to add the static helper methods as we do not have access to pickTasty method directly because we don't have access to the qctx where the transformation happens.

This seems to be an implementation detail, as a QuoteContext must be available in scope. Currently, in the implementation, it seems each quoted tree is associated with a `QuoteContext`, thus it's possible to get that in later compiler phases. It seems to me that polluting the API design with _accidental_ implementation details is not a good idea.

But there might be other aspects that I'm not aware of in the design. Therefore, feel free to choose whatever you think makes more sense technically.

Reply:
This design also generated reduces the code size of each quote by putting common code in those static methods. We do not want to generate the following code (or equivalent) at each place a quote is load.

```scala
val qctx = quoteContextWithCompilerInterface(summon[QuoteContext])
val tree = qctx.reflect.unpickleTerm(pickledQuote)
new scala.internal.quoted.Expr(tree, qctx.hashCode).asInstanceOf[Expr[T]]
```

Reply:
I would use this design as I can see some clear benefits for the optimization I have in mind. We can always evolve the API the way you said later if we need to do a major change.